[BPKR-64] Implement basic version of Bottom Sheet

fedetrim opened the pull request:
This PR introduces a new component for Backpack: Bottom Sheets.

+ [x] Check this if you have read and followed the [contributing guidelines](https://github.com/Skyscanner/backpack-ios/blob/master/CONTRIBUTING.md)

Remember to include the following changes:
+ [x] `UNRELEASED.md`
+ [x] `README.md`
+ [ ] Tests
+ [ ] Adding a component? Remember to expose it in the [main `Backpack.h` header file](https://github.com/Skyscanner/backpack-ios/tree/master/Backpack/Backpack.h)
+ [x] Docs (either update [backpack-docs](https://github.com/Skyscanner/backpack-docs) now, or create a follow up ticket)


_If you are curious about how we review, please read through the [code review guidelines](https://github.com/Skyscanner/backpack/blob/master/CODE_REVIEW_GUIDELINES.md)_

k0nserv commented on Example/Backpack/View Controllers/Bottom Sheet/BottomSheetBottomSectionVIewController.swift:
```diff
@@ -0,0 +1,34 @@
+
+}
+
+extension BottomSheetBottomSectionViewController: StoryboardInstantiable {
```
This is a super neat pattern, I like it :)

fedetrim commented on Backpack/BottomSheet/Classes/BackpackFloatingPanelController.swift:
```diff
@@ -0,0 +1,140 @@
+
+import FloatingPanel
+
+final class BackpackFloatingPanelController: FloatingPanelController {
```
I think there's a difference here. In this case, I'm using `Backpack` to distinguish it from the class we're inheriting from: `FloatingPanelController`. It's "our" internal subclass.
The `BPK` prefix, is there to solve an Obj-C limitation which is lack of namespacing. So we would use `BPK` in order to avoid collision with other's frameworks classes or even app code from our users.

I would only add `BPK` prefix to Obj-C code, and to Obj-C names for our Swift classes, like this:
```swift
@objc(BPKBackpackFloatingPanelController)
public final class BackpackFloatingPanelController: FloatingPanelController {
    ...
}
```

I don't think we need it in this case, as this class is internal and not public.
Anyways, if you still think we should add the prefix to Swift names, I can do it. Let me know.

k0nserv replied:
Generally we've avoided including the name `Backpack` or the prefix `BPK` in Swift type names because it doesn't seem very "Swifty"

fedetrim replied:
I agree, adding a prefix in Swift makes no sense. That's why I didn't go for `BPK`, because I don't want it to be a prefix. I just wanted it to be visible that we are using our own subclass and not the regular `FloatingPanelController`.

I could also just remove `Backpack` and redefine `FloatingPanelController` within our module.

What do you think we should do?

k0nserv replied:
Dunno maybe we should keep it if we know there will be conflicts immediately. Or rename it something like `BottomSheetController`? Regardless the Objective-C name shouldn't include the `BPK` prefix. @georgegillams what do you think?

georgegillams replied:
I think avoiding the conflict would be best if there is another obvious name we can use instead, but if not let's just be explicit and call it BackpackFloatingPanelController

fedetrim commented on Backpack/BottomSheet/Classes/BottomSheet.swift:
```diff
@@ -0,0 +1,142 @@
+public final class BottomSheet: NSObject {
+    
+    private struct Constants {
+        static let bottomSheetHeightInHalfPosition: CGFloat = 386.0
```
I'll check and get back to you. Nice catch by the way

fedetrim replied:
Looks good to me. What do you think?
![Simulator Screen Shot - iPhone SE - 2019-08-07 at 13 31 13](https://user-images.githubusercontent.com/3900360/62619708-b3984d00-b917-11e9-83c4-3a31c6d2596e.png)

k0nserv replied:
Yeah seems fine, thanks for checking :)

fedetrim commented on Backpack/BottomSheet/Classes/BottomSheet.swift:
```diff
@@ -0,0 +1,142 @@
+    
+    private lazy var floatingPanelController: BackpackFloatingPanelController = {
+        let panel = BackpackFloatingPanelController(delegate: self)
+        panel.surfaceView.cornerRadius = 24.0
```
Well, with our subclass I really didn't wanted to change any default functionality but rather to add something that we need and they don't provide: the ability to add a fixed bottom section that won't scroll.

So, setup is at the instantiation time rather than built-in in our subclass. Do you think I should move it?

k0nserv replied:
That sounds reasonable, let's leave it here

fedetrim commented on Backpack/BottomSheet/Classes/BottomSheet.swift:
```diff
@@ -0,0 +1,142 @@
+    public init(contentViewController: UIViewController, scrollViewToTrack: UIScrollView, bottomSectionViewController: UIViewController? = nil) {
+        super.init()
+        floatingPanelController.contentViewController = contentViewController
+        floatingPanelController.track(scrollView: scrollViewToTrack)
```
Right now, yes. If a scroll view isn't provided then we can assume that the content of the bottom sheet can fit in the screen without scrolling. So in that case, we might not want a hardcoded size (386.0) but rather want the bottom sheet to use the intrinsic size of the content.

There are also some other cases I can think of, but none are supported right now by our custom wrapper. By making the scroll view mandatory, we can postpone working on these for a later time when we actually need it.

Starting with a limited API and open it up later is no problem. But we would have a problem if we start we an open API (that makes the scroll view optional) and then we realize that it doesn't work as expected and we can't restrict the API because it would be a breaking change.

fedetrim replied:
What do you think?

k0nserv replied:
That sound sreasonable yeah

fedetrim commented:
Hey @k0nserv and @georgegillams, could you take a second look?
Just addressed PR comments: https://github.com/Skyscanner/backpack-ios/pull/366/commits/02350b9a05a2074a46a88c4ddfa924f3358d8a8e
And also updated README and UNRELEASED: https://github.com/Skyscanner/backpack-ios/pull/366/commits/cfe837d9ceedd04235e513dad8ae3cc136acad3b

fedetrim commented:
@k0nserv I was able to make Travis pass, but what about license/snyk and security/snyk? Why aren't those running?

georgegillams commented:
@fedetrim those seem to have just stopped running on all PR's over the last day or so. Will look into that!

fedetrim commented on Backpack/BottomSheet/README.md:
```diff
@@ -0,0 +1,66 @@
+
+### Bottom Sheet
+
+`Backpack/BottomSheet` contains the Backpack Bottom Sheet component in the class `BottomSheet`. The Bottom Sheet is a component for presenting a floating panel similar to the one used in Apple Maps, Stocks and other native Apple apps.
```
Applied your suggestion, I agree it's better. FYI, I was just following the README from the component Calendar. Maybe we should look into that one as well

georgegillams replied:
Will have a look at that!

fedetrim replied:
Oh no sorry, it wasn't Calendar, it was Button: https://github.com/Skyscanner/backpack-ios/blob/master/Backpack/Button/README.md

fedetrim commented on Backpack/BottomSheet/README.md:
````diff
@@ -0,0 +1,66 @@
+[bottomSheet presentInViewController:self animated:YES completion:nil];
+```
+
+With a fixed bottom section
````
I agree. Again, I followed the example in Button

fedetrim commented on Example/Backpack/View Controllers/Bottom Sheet/BottomSheetBottomSectionVIewController.swift:
```diff
@@ -0,0 +1,34 @@
+//
```
Good catch! Will do

fedetrim commented on .travis.yml:
```diff
@@ -35,7 +35,7 @@ before_script: 'if [[ "$MAIN_STAGE" = true ]]; then greenkeeper-lockfile-update;
   - npm run build
   - 'if [[ "$MAIN_STAGE" = true ]]; then ./scripts/check-pristine-state package-lock.json; fi'
-  - 'if [[ "$MAIN_STAGE" = true ]]; then Example/Pods/SwiftLint/swiftlint lint --strict --config Example/.swiftlint.yml; fi'
+  - 'if [[ "$MAIN_STAGE" = true ]]; then (cd Example && Pods/SwiftLint/swiftlint lint --strict --config .swiftlint.yml); fi'
```
What code is being excluded at the moment?

If I try your suggestion, then SwiftLint tries to run on Pods when running locally:
<img width="868" alt="Screen Shot 2019-08-12 at 17 53 59" src="https://user-images.githubusercontent.com/3900360/62878870-40814300-bd2a-11e9-913d-ce0f1d5bc307.png">
![Screen Shot 2019-08-12 at 17 52 09](https://user-images.githubusercontent.com/3900360/62878734-f39d6c80-bd29-11e9-8096-c0a770d3852f.png)

On the other hand, if I keep my changes, I see 51 files being linted locally and also the same 52 files being linted on the CI. I wonder what code you see being excluded?

georgegillams replied:
Well any swift code in `./Backpack` isn't being linted, (which at the moment is only your code)

fedetrim replied:
I see those being linted on the CI:
<img width="1018" alt="Screen Shot 2019-08-12 at 18 05 47" src="https://user-images.githubusercontent.com/3900360/62879630-d6699d80-bd2b-11e9-8710-80eb894dc733.png">

And also locally:
![Screen Shot 2019-08-12 at 18 06 30](https://user-images.githubusercontent.com/3900360/62879667-ef724e80-bd2b-11e9-84e5-63c4638e486f.png)

georgegillams replied:
That seems strange because if we cd into a different directory then it's not gonna have any knowledge of those files.

When I do this locally I can see swift files in Backpack being excluded due to the `cd` command:
```
 georgegillams@GEORGILL01M  ~/Documents/backpack-ios   master 13.0.0 ?  Example/Pods/SwiftLint/swiftlint lint --strict --config Example/.swiftlint.yml                                                                                                                                              ✔  10017  13:40:31
Loading configuration from 'Example/.swiftlint.yml'
Linting Swift files at paths
Linting 'ThemeRegistry.swift' (1/46)
Linting 'Settings.swift' (2/46)
Linting 'IconSwiftTest.swift' (3/46)
Linting 'TextViewUITest.swift' (4/46)
Linting 'TextFieldUITest.swift' (5/46)
Linting 'HorizontalNavigationUITest.swift' (6/46)
Linting 'ThemeHelpers.swift' (7/46)
Linting 'ChipUITest.swift' (8/46)
Linting 'DialogUITest.swift' (9/46)
Linting 'ExampleApp.swift' (10/46)
Linting 'TextViewsViewController.swift' (11/46)
Linting 'DialogSelectorViewController.swift' (12/46)
Linting 'CardsSelectorViewController.swift' (13/46)
Linting 'HorizontalNavViewController.swift' (14/46)
Linting 'ProgressBarViewController.swift' (15/46)
Linting 'LabelsSelectorTableViewController.swift' (16/46)
Linting 'TappableLinkLabelsViewController.swift' (17/46)
Linting 'GradientViewController.swift' (18/46)
Linting 'ChipsViewController.swift' (19/46)
Linting 'StarRatingsViewController.swift' (20/46)
Linting 'CardsViewController.swift' (21/46)
Linting 'SpinnersViewController.swift' (22/46)
Linting 'DialogViewController.swift' (23/46)
Linting 'HorizontalNavSelectorViewController.swift' (24/46)
Linting 'GradientSelectorTableViewController.swift' (25/46)
Linting 'ColorsViewController.swift' (26/46)
Linting 'SwitchesViewController.swift' (27/46)
Linting 'DividedCardsViewController.swift' (28/46)
Linting 'LabelsPerformanceViewController.swift' (29/46)
Linting 'LabelsViewController.swift' (30/46)
Linting 'TappableLinkLabelsSelectorViewController.swift' (31/46)
Linting 'TextFieldsViewController.swift' (32/46)
Linting 'CalendarViewController.swift' (33/46)
Linting 'BadgesViewController.swift' (34/46)
Linting 'LabelMultiFontStyleViewController.swift' (35/46)
Linting 'IconsViewController.swift' (36/46)
Linting 'SettingsViewController.swift' (37/46)
Linting 'ColorPreviewCollectionViewCell.swift' (38/46)
Linting 'ChipPreviewCollectionViewCell.swift' (39/46)
Linting 'IconsPreviewCollectionViewCell.swift' (40/46)
Linting 'UICollectionViewMasonryFlowLayout.swift' (41/46)
Linting 'PreviewCollectionViewHeader.swift' (42/46)
Linting 'BPKBadgeContainer.swift' (43/46)
Linting 'BPKTableViewSelectableCell.swift' (44/46)
Linting 'UIWindow+Utility.swift' (45/46)
Linting 'file-in-Backpack.swift' (46/46)
/Users/georgegillams/Documents/backpack-ios/Backpack/Badge/file-in-Backpack.swift:1:1: warning: Trailing Newline Violation: Files should have a single trailing newline. (trailing_newline)
Done linting! Found 1 violation, 0 serious in 46 files.
 georgegillams@GEORGILL01M  ~/Documents/backpack-ios   master 13.0.0 ?  (cd Example/ && Pods/SwiftLint/swiftlint lint --strict --config .swiftlint.yml)                                                                                                                                           3 ↵  10018  13:40:35
Reverting to nvm default version
Now using node v10.15.3 (npm v6.9.0)
Loading configuration from '.swiftlint.yml'
Linting Swift files at paths
Linting 'IconSwiftTest.swift' (1/45)
Linting 'Settings.swift' (2/45)
Linting 'ThemeRegistry.swift' (3/45)
Linting 'TextViewUITest.swift' (4/45)
Linting 'HorizontalNavigationUITest.swift' (5/45)
Linting 'TextFieldUITest.swift' (6/45)
Linting 'ThemeHelpers.swift' (7/45)
Linting 'ChipUITest.swift' (8/45)
Linting 'DialogUITest.swift' (9/45)
Linting 'ExampleApp.swift' (10/45)
Linting 'TextViewsViewController.swift' (11/45)
Linting 'DialogSelectorViewController.swift' (12/45)
Linting 'CardsSelectorViewController.swift' (13/45)
Linting 'HorizontalNavViewController.swift' (14/45)
Linting 'ProgressBarViewController.swift' (15/45)
Linting 'LabelsSelectorTableViewController.swift' (16/45)
Linting 'TappableLinkLabelsViewController.swift' (17/45)
Linting 'GradientViewController.swift' (18/45)
Linting 'ChipsViewController.swift' (19/45)
Linting 'StarRatingsViewController.swift' (20/45)
Linting 'CardsViewController.swift' (21/45)
Linting 'SpinnersViewController.swift' (22/45)
Linting 'DialogViewController.swift' (23/45)
Linting 'HorizontalNavSelectorViewController.swift' (24/45)
Linting 'GradientSelectorTableViewController.swift' (25/45)
Linting 'ColorsViewController.swift' (26/45)
Linting 'SwitchesViewController.swift' (27/45)
Linting 'DividedCardsViewController.swift' (28/45)
Linting 'LabelsPerformanceViewController.swift' (29/45)
Linting 'LabelsViewController.swift' (30/45)
Linting 'TappableLinkLabelsSelectorViewController.swift' (31/45)
Linting 'TextFieldsViewController.swift' (32/45)
Linting 'CalendarViewController.swift' (33/45)
Linting 'BadgesViewController.swift' (34/45)
Linting 'LabelMultiFontStyleViewController.swift' (35/45)
Linting 'IconsViewController.swift' (36/45)
Linting 'SettingsViewController.swift' (37/45)
Linting 'ColorPreviewCollectionViewCell.swift' (38/45)
Linting 'ChipPreviewCollectionViewCell.swift' (39/45)
Linting 'IconsPreviewCollectionViewCell.swift' (40/45)
Linting 'UICollectionViewMasonryFlowLayout.swift' (41/45)
Linting 'PreviewCollectionViewHeader.swift' (42/45)
Linting 'BPKBadgeContainer.swift' (43/45)
Linting 'BPKTableViewSelectableCell.swift' (44/45)
Linting 'UIWindow+Utility.swift' (45/45)
Done linting! Found 0 violations, 0 serious in 45 files.
```

georgegillams replied:
See how `file-in-Backpack.swift` only gets picked up on the first run?

fedetrim replied:
Second run is working fine for me:
```bash
FEDETRIM01M:backpack-ios federicotrimboli$ (cd Example/ && Pods/SwiftLint/swiftlint lint --strict --config .swiftlint.yml)
Loading configuration from '.swiftlint.yml'
Linting Swift files at paths 
Linting 'BottomSheet.swift' (1/51)
Linting 'BackpackFloatingPanelController.swift' (2/51)
Linting 'IconSwiftTest.swift' (3/51)
Linting 'Storyboards.swift' (4/51)
Linting 'ThemeRegistry.swift' (5/51)
Linting 'Settings.swift' (6/51)
Linting 'TextViewUITest.swift' (7/51)
Linting 'ThemeHelpers.swift' (8/51)
Linting 'HorizontalNavigationUITest.swift' (9/51)
Linting 'ChipUITest.swift' (10/51)
Linting 'TextFieldUITest.swift' (11/51)
Linting 'DialogUITest.swift' (12/51)
Linting 'ExampleApp.swift' (13/51)
Linting 'TextViewsViewController.swift' (14/51)
Linting 'DialogSelectorViewController.swift' (15/51)
Linting 'CardsSelectorViewController.swift' (16/51)
Linting 'HorizontalNavViewController.swift' (17/51)
Linting 'ProgressBarViewController.swift' (18/51)
Linting 'LabelsSelectorTableViewController.swift' (19/51)
Linting 'TappableLinkLabelsViewController.swift' (20/51)
Linting 'GradientViewController.swift' (21/51)
Linting 'ChipsViewController.swift' (22/51)
Linting 'StarRatingsViewController.swift' (23/51)
Linting 'CardsViewController.swift' (24/51)
Linting 'SpinnersViewController.swift' (25/51)
Linting 'DialogViewController.swift' (26/51)
Linting 'HorizontalNavSelectorViewController.swift' (27/51)
Linting 'GradientSelectorTableViewController.swift' (28/51)
Linting 'BottomSheetContentViewController.swift' (29/51)
Linting 'BottomSheetViewController.swift' (30/51)
Linting 'BottomSheetBottomSectionViewController.swift' (31/51)
Linting 'ColorsViewController.swift' (32/51)
Linting 'SwitchesViewController.swift' (33/51)
Linting 'DividedCardsViewController.swift' (34/51)
Linting 'LabelsPerformanceViewController.swift' (35/51)
Linting 'LabelsViewController.swift' (36/51)
Linting 'TappableLinkLabelsSelectorViewController.swift' (37/51)
Linting 'TextFieldsViewController.swift' (38/51)
Linting 'CalendarViewController.swift' (39/51)
Linting 'BadgesViewController.swift' (40/51)
Linting 'LabelMultiFontStyleViewController.swift' (41/51)
Linting 'IconsViewController.swift' (42/51)
Linting 'SettingsViewController.swift' (43/51)
Linting 'ColorPreviewCollectionViewCell.swift' (44/51)
Linting 'ChipPreviewCollectionViewCell.swift' (45/51)
Linting 'IconsPreviewCollectionViewCell.swift' (46/51)
Linting 'UICollectionViewMasonryFlowLayout.swift' (47/51)
Linting 'PreviewCollectionViewHeader.swift' (48/51)
Linting 'BPKBadgeContainer.swift' (49/51)
Linting 'BPKTableViewSelectableCell.swift' (50/51)
Linting 'UIWindow+Utility.swift' (51/51)
Done linting! Found 0 violations, 0 serious in 51 files.
Saving session...
...copying shared history...
...saving history...truncating history files...
...completed.
Deleting expired sessions...9 completed.

```

georgegillams replied:
Looks 💯 to me 👍

fedetrim commented:
@georgegillams 
>Also wondering how it might be possible for a consumer to wrap the View Controller, eg with the ThemeContainerController to support theming? I think to do that we might need to allow consumers to control the view hierarchy themselves...

Can't you just wrap `contentViewController` and `bottomSectionViewController`? Why do you need to wrap the bottom sheet itself? I think that asking for that is like asking for support to wrap a `UINavigationController` or `UITabBarController`. Those don't work when wrapped, either

k0nserv commented:
> Can't you just wrap contentViewController and bottomSectionViewController? Why do you need to wrap the bottom sheet itself? I think that asking for that is like asking for support to wrap a UINavigationController or UITabBarController. Those don't work when wrapped, either

`UINavigationController` and `UITabBarController` definitely work when wrapped.

fedetrim commented:
@k0nserv 

Yes, sorry, I expressed myself incorrectly. It *might work* (depending on what you're trying to do) but it's actually not supported:
>You can use the view controllers that the UIKit framework provides by themselves or in conjunction with other view controllers to create even more sophisticated interfaces. **When combining view controllers, however, the order of containment is important; only certain arrangements are valid. The order of containment, from child to parent, is as follows**:
>
>* Content view controllers, and container view controllers that have flexible bounds (such as the page view controller)
>* Navigation view controller
>* Tab bar controller
>* Split view controller

(source: https://developer.apple.com/library/archive/documentation/WindowsViews/Conceptual/ViewControllerCatalog/Chapters/CombiningViewControllers.html)

Container view controllers are at the top of the list, and not the other way around. It means that a navigation controller can contain a container view controller, but a container view controller cannot contain a navigation controller (unless it's a tab bar controller or a split view controller).

Given that, if you need to customize your content with a container, you would contain you **content view controllers**, not the containers.
In our case it's the same. Our bottom sheet is the container, and if we need customization, we should create wrappers for the content (`contentViewController` and `bottomSectionViewController`), not for the container.

Are we containing nav controllers, tab bar controllers or split view controllers with a custom container in any part of our code? If that's the case, in my opinion we should change it given it's not supported. It might break without notice in a future Xcode release.

-- 

Having said that, I don't think we this even if we wanted to. `FloatingPanel` framework ties the actual `FloatingPanelController` instance to the UIKit transitioning API, so that when doing this...
```swift
present(myFloatingPanelController, animated: true, completion: nil)
```
...the bottom sheet is presented with a custom navigation (same thing with dismissal) instead of being presented with the native modal transition.
If we wrapped the bottom sheet, and tried to present the container, custom transition wouldn't work. (And I think this is not a design error in `FloatingPanel` -- I wouldn't add support for containing a container without flexible bounds, either)

georgegillams commented:
Wrapping the `content` and `bottomSection` seem to work, except then I see this:
![image](https://user-images.githubusercontent.com/30267516/63096180-1e273980-bf65-11e9-93bb-e8fb4e16a786.png)


It's an issue with constraints but I'm struggling to work out how to fix it.

These are the changes I've made:
```
diff --git a/Example/Backpack/View Controllers/Bottom Sheet/BottomSheet.storyboard b/Example/Backpack/View Controllers/Bottom Sheet/BottomSheet.storyboard
index 0f5b90b..2518678 100644
--- a/Example/Backpack/View Controllers/Bottom Sheet/BottomSheet.storyboard	
+++ b/Example/Backpack/View Controllers/Bottom Sheet/BottomSheet.storyboard	
@@ -28,7 +28,7 @@
                                             <rect key="frame" x="0.0" y="0.0" width="414" height="43.5"/>
                                             <autoresizingMask key="autoresizingMask"/>
                                             <subviews>
-                                                <label opaque="NO" multipleTouchEnabled="YES" contentMode="left" insetsLayoutMarginsFromSafeArea="NO" text="Bottom Sheet with a scroll view" textAlignment="natural" lineBreakMode="tailTruncation" baselineAdjustment="alignBaselines" adjustsFontSizeToFit="NO" id="KWD-2d-n6w">
+                                                <label opaque="NO" multipleTouchEnabled="YES" contentMode="left" insetsLayoutMarginsFromSafeArea="NO" text="Bottom Sheet with a scroll view" textAlignment="natural" lineBreakMode="tailTruncation" baselineAdjustment="alignBaselines" adjustsFontSizeToFit="NO" id="KWD-2d-n6w" customClass="BPKLabel">
                                                     <rect key="frame" x="20" y="0.0" width="374" height="43.5"/>
                                                     <autoresizingMask key="autoresizingMask"/>
                                                     <fontDescription key="fontDescription" type="system" pointSize="17"/>
@@ -45,7 +45,7 @@
                                             <rect key="frame" x="0.0" y="0.0" width="414" height="43.5"/>
                                             <autoresizingMask key="autoresizingMask"/>
                                             <subviews>
-                                                <label opaque="NO" multipleTouchEnabled="YES" contentMode="left" insetsLayoutMarginsFromSafeArea="NO" text="Bottom Sheet with a bottom section" textAlignment="natural" lineBreakMode="tailTruncation" baselineAdjustment="alignBaselines" adjustsFontSizeToFit="NO" id="TL1-hE-1ep">
+                                                <label opaque="NO" multipleTouchEnabled="YES" contentMode="left" insetsLayoutMarginsFromSafeArea="NO" text="Bottom Sheet with a bottom section" textAlignment="natural" lineBreakMode="tailTruncation" baselineAdjustment="alignBaselines" adjustsFontSizeToFit="NO" id="TL1-hE-1ep" customClass="BPKLabel">
                                                     <rect key="frame" x="20" y="0.0" width="374" height="43.5"/>
                                                     <autoresizingMask key="autoresizingMask"/>
                                                     <fontDescription key="fontDescription" type="system" pointSize="17"/>
@@ -91,7 +91,7 @@
                                             <rect key="frame" x="0.0" y="0.0" width="414" height="43.5"/>
                                             <autoresizingMask key="autoresizingMask"/>
                                             <subviews>
-                                                <label opaque="NO" multipleTouchEnabled="YES" contentMode="left" insetsLayoutMarginsFromSafeArea="NO" text="First Row" textAlignment="natural" lineBreakMode="tailTruncation" baselineAdjustment="alignBaselines" adjustsFontSizeToFit="NO" id="Sh0-gy-djR">
+                                                <label opaque="NO" multipleTouchEnabled="YES" contentMode="left" insetsLayoutMarginsFromSafeArea="NO" text="First Row" textAlignment="natural" lineBreakMode="tailTruncation" baselineAdjustment="alignBaselines" adjustsFontSizeToFit="NO" id="Sh0-gy-djR" customClass="BPKLabel">
                                                     <rect key="frame" x="20" y="0.0" width="374" height="43.5"/>
                                                     <autoresizingMask key="autoresizingMask"/>
                                                     <fontDescription key="fontDescription" type="system" pointSize="17"/>
@@ -108,7 +108,7 @@
                                             <rect key="frame" x="0.0" y="0.0" width="414" height="43.5"/>
                                             <autoresizingMask key="autoresizingMask"/>
                                             <subviews>
-                                                <label opaque="NO" multipleTouchEnabled="YES" contentMode="left" insetsLayoutMarginsFromSafeArea="NO" text="Second Row" textAlignment="natural" lineBreakMode="tailTruncation" baselineAdjustment="alignBaselines" adjustsFontSizeToFit="NO" id="Lee-UN-SPE">
+                                                <label opaque="NO" multipleTouchEnabled="YES" contentMode="left" insetsLayoutMarginsFromSafeArea="NO" text="Second Row" textAlignment="natural" lineBreakMode="tailTruncation" baselineAdjustment="alignBaselines" adjustsFontSizeToFit="NO" id="Lee-UN-SPE" customClass="BPKLabel">
                                                     <rect key="frame" x="20" y="0.0" width="374" height="43.5"/>
                                                     <autoresizingMask key="autoresizingMask"/>
                                                     <fontDescription key="fontDescription" type="system" pointSize="17"/>
@@ -125,7 +125,7 @@
                                             <rect key="frame" x="0.0" y="0.0" width="414" height="43.5"/>
                                             <autoresizingMask key="autoresizingMask"/>
                                             <subviews>
-                                                <label opaque="NO" multipleTouchEnabled="YES" contentMode="left" insetsLayoutMarginsFromSafeArea="NO" text="Third Row" textAlignment="natural" lineBreakMode="tailTruncation" baselineAdjustment="alignBaselines" adjustsFontSizeToFit="NO" id="neY-NZ-Ox5">
+                                                <label opaque="NO" multipleTouchEnabled="YES" contentMode="left" insetsLayoutMarginsFromSafeArea="NO" text="Third Row" textAlignment="natural" lineBreakMode="tailTruncation" baselineAdjustment="alignBaselines" adjustsFontSizeToFit="NO" id="neY-NZ-Ox5" customClass="BPKLabel">
                                                     <rect key="frame" x="20" y="0.0" width="374" height="43.5"/>
                                                     <autoresizingMask key="autoresizingMask"/>
                                                     <fontDescription key="fontDescription" type="system" pointSize="17"/>
@@ -146,7 +146,7 @@
                                             <rect key="frame" x="0.0" y="0.0" width="414" height="43.5"/>
                                             <autoresizingMask key="autoresizingMask"/>
                                             <subviews>
-                                                <label opaque="NO" multipleTouchEnabled="YES" contentMode="left" insetsLayoutMarginsFromSafeArea="NO" text="First Row" textAlignment="natural" lineBreakMode="tailTruncation" baselineAdjustment="alignBaselines" adjustsFontSizeToFit="NO" id="pZT-4w-AaB">
+                                                <label opaque="NO" multipleTouchEnabled="YES" contentMode="left" insetsLayoutMarginsFromSafeArea="NO" text="First Row" textAlignment="natural" lineBreakMode="tailTruncation" baselineAdjustment="alignBaselines" adjustsFontSizeToFit="NO" id="pZT-4w-AaB" customClass="BPKLabel">
                                                     <rect key="frame" x="20" y="0.0" width="374" height="43.5"/>
                                                     <autoresizingMask key="autoresizingMask"/>
                                                     <fontDescription key="fontDescription" type="system" pointSize="17"/>
@@ -163,7 +163,7 @@
                                             <rect key="frame" x="0.0" y="0.0" width="414" height="43.5"/>
                                             <autoresizingMask key="autoresizingMask"/>
                                             <subviews>
-                                                <label opaque="NO" multipleTouchEnabled="YES" contentMode="left" insetsLayoutMarginsFromSafeArea="NO" text="Second Row" textAlignment="natural" lineBreakMode="tailTruncation" baselineAdjustment="alignBaselines" adjustsFontSizeToFit="NO" id="f7d-GU-dCM">
+                                                <label opaque="NO" multipleTouchEnabled="YES" contentMode="left" insetsLayoutMarginsFromSafeArea="NO" text="Second Row" textAlignment="natural" lineBreakMode="tailTruncation" baselineAdjustment="alignBaselines" adjustsFontSizeToFit="NO" id="f7d-GU-dCM" customClass="BPKLabel">
                                                     <rect key="frame" x="20" y="0.0" width="374" height="43.5"/>
                                                     <autoresizingMask key="autoresizingMask"/>
                                                     <fontDescription key="fontDescription" type="system" pointSize="17"/>
@@ -180,7 +180,7 @@
                                             <rect key="frame" x="0.0" y="0.0" width="414" height="43.5"/>
                                             <autoresizingMask key="autoresizingMask"/>
                                             <subviews>
-                                                <label opaque="NO" multipleTouchEnabled="YES" contentMode="left" insetsLayoutMarginsFromSafeArea="NO" text="Third Row" textAlignment="natural" lineBreakMode="tailTruncation" baselineAdjustment="alignBaselines" adjustsFontSizeToFit="NO" id="KcR-LK-omM">
+                                                <label opaque="NO" multipleTouchEnabled="YES" contentMode="left" insetsLayoutMarginsFromSafeArea="NO" text="Third Row" textAlignment="natural" lineBreakMode="tailTruncation" baselineAdjustment="alignBaselines" adjustsFontSizeToFit="NO" id="KcR-LK-omM" customClass="BPKLabel">
                                                     <rect key="frame" x="20" y="0.0" width="374" height="43.5"/>
                                                     <autoresizingMask key="autoresizingMask"/>
                                                     <fontDescription key="fontDescription" type="system" pointSize="17"/>
@@ -201,7 +201,7 @@
                                             <rect key="frame" x="0.0" y="0.0" width="414" height="43.5"/>
                                             <autoresizingMask key="autoresizingMask"/>
                                             <subviews>
-                                                <label opaque="NO" multipleTouchEnabled="YES" contentMode="left" insetsLayoutMarginsFromSafeArea="NO" text="First Row" textAlignment="natural" lineBreakMode="tailTruncation" baselineAdjustment="alignBaselines" adjustsFontSizeToFit="NO" id="XeX-ZH-KSk">
+                                                <label opaque="NO" multipleTouchEnabled="YES" contentMode="left" insetsLayoutMarginsFromSafeArea="NO" text="First Row" textAlignment="natural" lineBreakMode="tailTruncation" baselineAdjustment="alignBaselines" adjustsFontSizeToFit="NO" id="XeX-ZH-KSk" customClass="BPKLabel">
                                                     <rect key="frame" x="20" y="0.0" width="374" height="43.5"/>
                                                     <autoresizingMask key="autoresizingMask"/>
                                                     <fontDescription key="fontDescription" type="system" pointSize="17"/>
@@ -218,7 +218,7 @@
                                             <rect key="frame" x="0.0" y="0.0" width="414" height="43.5"/>
                                             <autoresizingMask key="autoresizingMask"/>
                                             <subviews>
-                                                <label opaque="NO" multipleTouchEnabled="YES" contentMode="left" insetsLayoutMarginsFromSafeArea="NO" text="Second Row" textAlignment="natural" lineBreakMode="tailTruncation" baselineAdjustment="alignBaselines" adjustsFontSizeToFit="NO" id="VSL-zH-Ke0">
+                                                <label opaque="NO" multipleTouchEnabled="YES" contentMode="left" insetsLayoutMarginsFromSafeArea="NO" text="Second Row" textAlignment="natural" lineBreakMode="tailTruncation" baselineAdjustment="alignBaselines" adjustsFontSizeToFit="NO" id="VSL-zH-Ke0" customClass="BPKLabel">
                                                     <rect key="frame" x="20" y="0.0" width="374" height="43.5"/>
                                                     <autoresizingMask key="autoresizingMask"/>
                                                     <fontDescription key="fontDescription" type="system" pointSize="17"/>
@@ -235,7 +235,7 @@
                                             <rect key="frame" x="0.0" y="0.0" width="414" height="43.5"/>
                                             <autoresizingMask key="autoresizingMask"/>
                                             <subviews>
-                                                <label opaque="NO" multipleTouchEnabled="YES" contentMode="left" insetsLayoutMarginsFromSafeArea="NO" text="Third Row" textAlignment="natural" lineBreakMode="tailTruncation" baselineAdjustment="alignBaselines" adjustsFontSizeToFit="NO" id="FGI-0B-EoC">
+                                                <label opaque="NO" multipleTouchEnabled="YES" contentMode="left" insetsLayoutMarginsFromSafeArea="NO" text="Third Row" textAlignment="natural" lineBreakMode="tailTruncation" baselineAdjustment="alignBaselines" adjustsFontSizeToFit="NO" id="FGI-0B-EoC" customClass="BPKLabel">
                                                     <rect key="frame" x="20" y="0.0" width="374" height="43.5"/>
                                                     <autoresizingMask key="autoresizingMask"/>
                                                     <fontDescription key="fontDescription" type="system" pointSize="17"/>
@@ -256,7 +256,7 @@
                                             <rect key="frame" x="0.0" y="0.0" width="414" height="43.5"/>
                                             <autoresizingMask key="autoresizingMask"/>
                                             <subviews>
-                                                <label opaque="NO" multipleTouchEnabled="YES" contentMode="left" insetsLayoutMarginsFromSafeArea="NO" text="First Row" textAlignment="natural" lineBreakMode="tailTruncation" baselineAdjustment="alignBaselines" adjustsFontSizeToFit="NO" id="vjS-pC-XHw">
+                                                <label opaque="NO" multipleTouchEnabled="YES" contentMode="left" insetsLayoutMarginsFromSafeArea="NO" text="First Row" textAlignment="natural" lineBreakMode="tailTruncation" baselineAdjustment="alignBaselines" adjustsFontSizeToFit="NO" id="vjS-pC-XHw" customClass="BPKLabel">
                                                     <rect key="frame" x="20" y="0.0" width="374" height="43.5"/>
                                                     <autoresizingMask key="autoresizingMask"/>
                                                     <fontDescription key="fontDescription" type="system" pointSize="17"/>
@@ -273,7 +273,7 @@
                                             <rect key="frame" x="0.0" y="0.0" width="414" height="43.5"/>
                                             <autoresizingMask key="autoresizingMask"/>
                                             <subviews>
-                                                <label opaque="NO" multipleTouchEnabled="YES" contentMode="left" insetsLayoutMarginsFromSafeArea="NO" text="Second Row" textAlignment="natural" lineBreakMode="tailTruncation" baselineAdjustment="alignBaselines" adjustsFontSizeToFit="NO" id="vbx-W3-zkV">
+                                                <label opaque="NO" multipleTouchEnabled="YES" contentMode="left" insetsLayoutMarginsFromSafeArea="NO" text="Second Row" textAlignment="natural" lineBreakMode="tailTruncation" baselineAdjustment="alignBaselines" adjustsFontSizeToFit="NO" id="vbx-W3-zkV" customClass="BPKLabel">
                                                     <rect key="frame" x="20" y="0.0" width="374" height="43.5"/>
                                                     <autoresizingMask key="autoresizingMask"/>
                                                     <fontDescription key="fontDescription" type="system" pointSize="17"/>
@@ -290,7 +290,7 @@
                                             <rect key="frame" x="0.0" y="0.0" width="414" height="43.5"/>
                                             <autoresizingMask key="autoresizingMask"/>
                                             <subviews>
-                                                <label opaque="NO" multipleTouchEnabled="YES" contentMode="left" insetsLayoutMarginsFromSafeArea="NO" text="Third Row" textAlignment="natural" lineBreakMode="tailTruncation" baselineAdjustment="alignBaselines" adjustsFontSizeToFit="NO" id="0Cm-Db-rBD">
+                                                <label opaque="NO" multipleTouchEnabled="YES" contentMode="left" insetsLayoutMarginsFromSafeArea="NO" text="Third Row" textAlignment="natural" lineBreakMode="tailTruncation" baselineAdjustment="alignBaselines" adjustsFontSizeToFit="NO" id="0Cm-Db-rBD" customClass="BPKLabel">
                                                     <rect key="frame" x="20" y="0.0" width="374" height="43.5"/>
                                                     <autoresizingMask key="autoresizingMask"/>
                                                     <fontDescription key="fontDescription" type="system" pointSize="17"/>
@@ -311,7 +311,7 @@
                                             <rect key="frame" x="0.0" y="0.0" width="414" height="43.5"/>
                                             <autoresizingMask key="autoresizingMask"/>
                                             <subviews>
-                                                <label opaque="NO" multipleTouchEnabled="YES" contentMode="left" insetsLayoutMarginsFromSafeArea="NO" text="First Row" textAlignment="natural" lineBreakMode="tailTruncation" baselineAdjustment="alignBaselines" adjustsFontSizeToFit="NO" id="GQK-uy-kGe">
+                                                <label opaque="NO" multipleTouchEnabled="YES" contentMode="left" insetsLayoutMarginsFromSafeArea="NO" text="First Row" textAlignment="natural" lineBreakMode="tailTruncation" baselineAdjustment="alignBaselines" adjustsFontSizeToFit="NO" id="GQK-uy-kGe" customClass="BPKLabel">
                                                     <rect key="frame" x="20" y="0.0" width="374" height="43.5"/>
                                                     <autoresizingMask key="autoresizingMask"/>
                                                     <fontDescription key="fontDescription" type="system" pointSize="17"/>
@@ -328,7 +328,7 @@
                                             <rect key="frame" x="0.0" y="0.0" width="414" height="43.5"/>
                                             <autoresizingMask key="autoresizingMask"/>
                                             <subviews>
-                                                <label opaque="NO" multipleTouchEnabled="YES" contentMode="left" insetsLayoutMarginsFromSafeArea="NO" text="Second Row" textAlignment="natural" lineBreakMode="tailTruncation" baselineAdjustment="alignBaselines" adjustsFontSizeToFit="NO" id="t2h-ay-M57">
+                                                <label opaque="NO" multipleTouchEnabled="YES" contentMode="left" insetsLayoutMarginsFromSafeArea="NO" text="Second Row" textAlignment="natural" lineBreakMode="tailTruncation" baselineAdjustment="alignBaselines" adjustsFontSizeToFit="NO" id="t2h-ay-M57" customClass="BPKLabel">
                                                     <rect key="frame" x="20" y="0.0" width="374" height="43.5"/>
                                                     <autoresizingMask key="autoresizingMask"/>
                                                     <fontDescription key="fontDescription" type="system" pointSize="17"/>
@@ -345,7 +345,7 @@
                                             <rect key="frame" x="0.0" y="0.0" width="414" height="43.5"/>
                                             <autoresizingMask key="autoresizingMask"/>
                                             <subviews>
-                                                <label opaque="NO" multipleTouchEnabled="YES" contentMode="left" insetsLayoutMarginsFromSafeArea="NO" text="Third Row" textAlignment="natural" lineBreakMode="tailTruncation" baselineAdjustment="alignBaselines" adjustsFontSizeToFit="NO" id="sEa-5O-g4s">
+                                                <label opaque="NO" multipleTouchEnabled="YES" contentMode="left" insetsLayoutMarginsFromSafeArea="NO" text="Third Row" textAlignment="natural" lineBreakMode="tailTruncation" baselineAdjustment="alignBaselines" adjustsFontSizeToFit="NO" id="sEa-5O-g4s" customClass="BPKLabel">
                                                     <rect key="frame" x="20" y="0.0" width="374" height="43.5"/>
                                                     <autoresizingMask key="autoresizingMask"/>
                                                     <fontDescription key="fontDescription" type="system" pointSize="17"/>
@@ -366,7 +366,7 @@
                                             <rect key="frame" x="0.0" y="0.0" width="414" height="43.5"/>
                                             <autoresizingMask key="autoresizingMask"/>
                                             <subviews>
-                                                <label opaque="NO" multipleTouchEnabled="YES" contentMode="left" insetsLayoutMarginsFromSafeArea="NO" text="First Row" textAlignment="natural" lineBreakMode="tailTruncation" baselineAdjustment="alignBaselines" adjustsFontSizeToFit="NO" id="ekg-Nu-Gwu">
+                                                <label opaque="NO" multipleTouchEnabled="YES" contentMode="left" insetsLayoutMarginsFromSafeArea="NO" text="First Row" textAlignment="natural" lineBreakMode="tailTruncation" baselineAdjustment="alignBaselines" adjustsFontSizeToFit="NO" id="ekg-Nu-Gwu" customClass="BPKLabel">
                                                     <rect key="frame" x="15" y="0.0" width="384" height="43.5"/>
                                                     <autoresizingMask key="autoresizingMask"/>
                                                     <fontDescription key="fontDescription" type="system" pointSize="17"/>
@@ -383,7 +383,7 @@
                                             <rect key="frame" x="0.0" y="0.0" width="414" height="43.5"/>
                                             <autoresizingMask key="autoresizingMask"/>
                                             <subviews>
-                                                <label opaque="NO" multipleTouchEnabled="YES" contentMode="left" insetsLayoutMarginsFromSafeArea="NO" text="Second Row" textAlignment="natural" lineBreakMode="tailTruncation" baselineAdjustment="alignBaselines" adjustsFontSizeToFit="NO" id="jDj-Jr-sh8">
+                                                <label opaque="NO" multipleTouchEnabled="YES" contentMode="left" insetsLayoutMarginsFromSafeArea="NO" text="Second Row" textAlignment="natural" lineBreakMode="tailTruncation" baselineAdjustment="alignBaselines" adjustsFontSizeToFit="NO" id="jDj-Jr-sh8" customClass="BPKLabel">
                                                     <rect key="frame" x="15" y="0.0" width="384" height="43.5"/>
                                                     <autoresizingMask key="autoresizingMask"/>
                                                     <fontDescription key="fontDescription" type="system" pointSize="17"/>
@@ -400,7 +400,7 @@
                                             <rect key="frame" x="0.0" y="0.0" width="414" height="43.5"/>
                                             <autoresizingMask key="autoresizingMask"/>
                                             <subviews>
-                                                <label opaque="NO" multipleTouchEnabled="YES" contentMode="left" insetsLayoutMarginsFromSafeArea="NO" text="Third Row" textAlignment="natural" lineBreakMode="tailTruncation" baselineAdjustment="alignBaselines" adjustsFontSizeToFit="NO" id="vMx-kg-8Qb">
+                                                <label opaque="NO" multipleTouchEnabled="YES" contentMode="left" insetsLayoutMarginsFromSafeArea="NO" text="Third Row" textAlignment="natural" lineBreakMode="tailTruncation" baselineAdjustment="alignBaselines" adjustsFontSizeToFit="NO" id="vMx-kg-8Qb" customClass="BPKLabel">
                                                     <rect key="frame" x="15" y="0.0" width="384" height="43.5"/>
                                                     <autoresizingMask key="autoresizingMask"/>
                                                     <fontDescription key="fontDescription" type="system" pointSize="17"/>
diff --git a/Example/Backpack/View Controllers/Bottom Sheet/BottomSheetViewController.swift b/Example/Backpack/View Controllers/Bottom Sheet/BottomSheetViewController.swift
index 3323d53..21b6456 100644
--- a/Example/Backpack/View Controllers/Bottom Sheet/BottomSheetViewController.swift	
+++ b/Example/Backpack/View Controllers/Bottom Sheet/BottomSheetViewController.swift	
@@ -30,6 +30,11 @@ final class BottomSheetViewController: UITableViewController {
 extension BottomSheetViewController {
 
     override func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
+        guard let rootViewController =  UIApplication.shared.keyWindow?.rootViewController as?
+            ThemeContainerController else {
+                return
+        }
+
         tableView.deselectRow(at: indexPath, animated: true)
 
         let cell = tableView.cellForRow(at: indexPath)
@@ -37,16 +42,23 @@ extension BottomSheetViewController {
         case scrollViewBottomSheet:
             guard let content = BottomSheetContentViewController.make() else { return }
 
-            let sheet = BottomSheet(contentViewController: content,
+            let wrappedContent = rootViewController.createIdenticalContainerController(forRootController: content)
+
+            let sheet = BottomSheet(contentViewController: wrappedContent,
                                     scrollViewToTrack: content.tableView)
             sheet.present(in: self, animated: true, completion: nil)
         case bottomSectionBottomSheet:
             guard let content = BottomSheetContentViewController.make(),
                 let bottomSection = BottomSheetBottomSectionViewController.make() else { return }
 
-            let sheet = BottomSheet(contentViewController: content,
+            let wrappedContent = rootViewController.createIdenticalContainerController(forRootController: content)
+            let wrappedBottomSection = rootViewController.createIdenticalContainerController(
+                forRootController: bottomSection
+            )
+
+            let sheet = BottomSheet(contentViewController: wrappedContent,
                                     scrollViewToTrack: content.tableView,
-                                    bottomSectionViewController: bottomSection)
+                                    bottomSectionViewController: wrappedBottomSection)
 
             bottomSection.dismissClosure = {
                 sheet.viewControllerToPresent.dismiss(animated: true, completion: nil)
```

fedetrim commented:
@georgegillams let me try it out and get back to you

k0nserv commented:
Okay fair points about the containment issues let's support containing the inner controllers then. As far as `BPKThemeContainerController` goes yes we do use it to contain `UINavigationBarController` and `UITabBarController`, the docs you linked are new to me so thanks for highlighting that. I'm not too concerned about it though since: 

1. We just give the `UINavigationController` the full window size without constraining the size.
2. It's actually really common in community libraries(drawers for example are almost always implemented like this).

fedetrim commented:
@georgegillams 
Just pushed the fix: https://github.com/Skyscanner/backpack-ios/pull/366/commits/a3e8656b48cfcaa1a3d18a1735d83e0e4d45b552
If you pull it, your local changes should work.

@k0nserv 
If it really is common in community libraries, then I suppose Apple will take care not to break it. Once an unsupported practice becomes mainstream, then they are forced to support it, even if they never comment on it publicly. I'd try to avoid using it in new code, but I agree it's not worth the refactor in order to remove it from legacy code

georgegillams commented:
Definitely working a lot better now. Any idea why that dark shadow is appearing?
<img width="1060" alt="image" src="https://user-images.githubusercontent.com/30267516/63266686-62c61400-c288-11e9-9d8e-03a73c2a4ebe.png">

fedetrim commented:
@georgegillams yes, the shadow is automatically added by our Bottom Sheet component. The color of the shadow depends on the `backgroundColor` of the `bottomSectionViewController`. In this case, it's the background color of the container, which I suppose is `.black`.

If you set the `backgroundColor` of the container to `.clear`, the shadow should disappear. If you set it to `.white`, the shadow color should change to white.

georgegillams commented:
Awesome - do you mind apply this diff using `git apply`, and then it should be all 💯 
[git_diff.txt](https://github.com/Skyscanner/backpack-ios/files/3515991/git_diff.txt)

fedetrim commented:
@georgegillams just pushed, thanks again for reviewing 🙂
